### DISCUSSION TITLE
Coplanar union and plane cut

### DISCUSSION THREAD

**Discussion by hrgdavor** [elalish/manifold#287] [Ideas]
one of common ways I sometimes model stuff for 3d printing is stacking different sections and then making a union. It helps me to break complex model into more manageable parts.

Since manifold is currently very actively developed, I am thinking that there are some cases where boolean ops could be optimized by having some known constraint that reduces amount of work for the alg.

Simple case to demonstrate the concept would be two cylinder one on top of the other.
![image](https://user-images.githubusercontent.com/2480762/202902562-2f90e5cf-761e-4ae9-bfa1-938f43eb0957.png)

My experience is from jscad csg.js bsp tree where space gets cut into a soup, and I lack skill to implement it there. 

I am sure there are more complex cases where there would be real benefit to performance.

Maybe manifold by design handles such cases much better, I just do not know. I want to pick your brain a little on this topic.

Bonus question: is there an operation in manifold that cuts a mesh in some plane ?

**Comment by pca006132** [comment]
>  is there an operation in manifold that cuts a mesh in some plane ?

We do:
https://github.com/elalish/manifold/blob/87059508e76ae097baddde2256c8bb805859bf27/src/manifold/include/manifold.h#L168

But this is so rarely used that we forgot to export it in our bindings :P
Perhaps we should fix this.

**Reply by hrgdavor** [reply]
that is very nice :) ... feels silly when I want to do that in jscad, I need to make a large enough cube and subtract it. 🤦

**Reply by elalish** [reply]
Yeah, turns out it's doing exactly that under the hood, but I made this convenience function exactly because I needed it so often. As for performance, I'd say we're already decently optimized and the real beauty of particularly the boolean code is that it avoids branching and special-casing. That's what keeps it reliable.

**Comment by pca006132** [comment]
And for optimizing boolean operations with constraints: currently we will lazily perform the actual boolean operation, most often the operations will just create some operation tree, which can be optimized by reordering, using compose instead of union for disjoint parts, etc. Although we did not really look deep into better heuristics in that area, the optimizations currently implemented is quite trivial.
